GRPA-3100: Consolidate storage configuration and convert to Butane (#33337)
openshift/openshift-docs: modules/installation-special-config-storage.adoc

+
[IMPORTANT]
====
Some methods for configuring static IPs do not affect the initramfs after the first boot, and so will not work with Tang encryption.  These include the coreos-installer `--copy-network` option, and adding `ip=` arguments to the kernel command line of the live ISO or PXE image during installation.  Incorrect static IP configuration causes the second boot of the node to fail.
====

. Change to the directory that contains the installation program and generate the Kubernetes manifests for the cluster:
+
[source,terminal]
----
$ ./openshift-install create manifests --dir=<installation_directory>
----

. Create a Butane config configuring disk encryption and/or mirroring. For example, to configure storage for worker nodes, create `$HOME/clusterconfig/worker-storage.bu`:
+
[source,yaml]
.Butane config example
----
variant: openshift
version: 4.8.0
metadata:
  name: worker-storage <1>
  labels:
    machineconfiguration.openshift.io/role: worker <1>

> boot_device:

Reviewer: Should probably include the `layout` key as we've mentioned above that this feature is available on `x86_64 nodes booted with BIOS or UEFI and on ppc64le nodes.`

Author: Good catch.  Updated.

During {product-title} installation on master and worker nodes, you can enable mirroring of the boot disk to two or more redundant storage devices. A node continues to function after storage device failure as long as one device remains available.

Mirroring does not support replacement of a failed disk. To restore the mirror to a pristine, non-degraded state, reprovision the node.

[NOTE]
====
Mirroring is available only for user-provisioned infrastructure deployments on {op-system} systems. Mirroring support is available on x86_64 nodes booted with BIOS or UEFI and on ppc64le nodes.
====

[id="installation-special-config-storage-procedure_{context}"]
== Configuring disk encryption or mirroring

.Procedure

To enable boot disk encryption or mirroring during {product-title} deployment:

. If you want to use TPM v2 to encrypt your cluster, check to see if TPM v2 encryption needs to be enabled in the BIOS on each node. This is required on most Dell systems. Check the manual for your computer.

. If you want to use Tang to encrypt your cluster, follow these preparatory steps:

.. Access a Red Hat Enterprise Linux server from which you can configure the encryption settings.

> .. Set up or access an existing Tang server. See link:https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/security_hardening/configuring-automated-unlocking-of-encrypted-volumes-using-policy-based-decryption_security-hardening#network-bound-disk-encryption_configuring-automated-unlocking-of-encrypted-volumes-using-policy-based-decryption[Network-bound disk encryption] for instructions. See link:https://youtu.be/2uLKvB8Z5D0[Securing Automated Decryption New Cryptography and Techniques] for a presentation on Tang.

Author: This was carried over verbatim from the [existing docs](https://docs.openshift.com/container-platform/4.7/installing/install_config/installing-customizing.html#installation-special-config-encrypt-disk-tang_installing-customizing), originally added in #19144.  I don't have an opinion on keeping the link; lmk if you'd like it removed.

> .. Set up or access an existing Tang server. See link:https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/security_hardening/configuring-automated-unlocking-of-encrypted-volumes-using-policy-based-decryption_security-hardening#network-bound-disk-encryption_configuring-automated-unlocking-of-encrypted-volumes-using-policy-based-decryption[Network-bound disk encryption] for instructions. See link:https://youtu.be/2uLKvB8Z5D0[Securing Automated Decryption New Cryptography and Techniques] for a presentation on Tang.

Reviewer 2: In that case, let's keep it! 😆

.. Install the `clevis` package, if it is not already installed:
+
[source,terminal]
----
$ sudo yum install clevis -y
----

.. Generate a thumbprint for the Tang server.

... In the following command, replace the value of `url` with the Tang server URL:
+
[source,terminal]
----
$ echo nifty random wordwords | clevis-encrypt-tang '{"url":"https://tang.example.org"}'
----
+
.Example output
[source,terminal]
----
The advertisement contains the following signing keys:

PLjNyRdGw03zlRoGjQYMahSZGu9
----

... When the `Do you wish to trust these keys? [ynYN]` prompt displays, type `Y`, and the thumbprint is displayed:
+
.Example output
[source,terminal]
----
eyJhbmc3SlRyMXpPenc3ajhEQ01tZVJiTi1oM...
----

.. If nodes are configured with static IP addressing, ensure you use the coreos-installer `--append-karg` option when installing {op-system} nodes to set the IP address of the installed system.  Append the `ip=` and other arguments needed for your network.
+
[IMPORTANT]
====

> Some methods for configuring static IPs do not affect the initramfs after the first boot, and so will not work with Tang encryption.  These include the coreos-installer `--copy-network` option, and adding `ip=` arguments to the kernel command line of the live ISO or PXE image during installation.  Incorrect static IP configuration causes the second boot of the node to fail.

Reviewer 2: ```suggestion
Some methods for configuring static IPs do not affect the initramfs after the first boot and will not work with Tang encryption. These include the `coreos-installer` `--copy-network` option, as well as adding `ip=` arguments to the kernel command line of the live ISO or PXE image during installation. Incorrect static IP configuration causes the second boot of the node to fail.
```

^Tiny style tweaks

====

. Change to the directory that contains the installation program and generate the Kubernetes manifests for the cluster:
+
[source,terminal]
----
$ ./openshift-install create manifests --dir=<installation_directory>
----

> . Create a Butane config configuring disk encryption and/or mirroring. For example, to configure storage for worker nodes, create `$HOME/clusterconfig/worker-storage.bu`:

Reviewer 2: ```suggestion
. Create a Butane config that configures disk encryption, mirroring, or both. For example, to configure storage for worker nodes, create a `$HOME/clusterconfig/worker-storage.bu` file:
```
^For clarity, but please ignore if my changes are incorrect.

+
[source,yaml]
.Butane config example
----
variant: openshift
version: 4.8.0
metadata:
  name: worker-storage <1>
  labels:
    machineconfiguration.openshift.io/role: worker <1>
boot_device:
  layout: x86_64 <2>
  luks: <3>
    tpm2: true <4>
    tang: <5>
      - url: https://tang.example.com
        thumbprint: eyJhbmc3SlRyMXpPenc3ajhEQ01tZVJiTi1oM...
    threshold: 1 <6>
  mirror: <7>
    devices: <8>
      - /dev/sda
      - /dev/sdb
openshift:
  fips: true <9>
----
+

> <1> On master nodes, substitute `master` for `worker` in both of these locations.

Reviewer 2: Similar comments:
"On control plane nodes, substitute `master` (or control plane) for `worker` in both of these locations.

<2> On ppc64le nodes, set this field to `ppc64le`.  On all other nodes, this field can be omitted.
<3> Include this section if you want to encrypt the root filesystem. See "About disk encryption" for more details.
<4> Include this field if you want to use a Trusted Platform Module (TPM) to encrypt the root filesystem.

> <5> Include this section if you want to use a Tang server. To obtain the server URL and thumbprint, follow the instructions above.

Reviewer 2: s/the instructions above./the preceding instructions./ (IBMSG again)

<6> Include this directive to use Shamir's Secret Sharing (SSS) to require the specified number of Tang servers, or one less than the specified number of servers plus a TPM, to unlock the disk.
<7> Include this section if you want to mirror the boot disk. See "About disk mirroring" for more details.
<8> List all disk devices that should be included in the boot disk mirror, including the disk that {op-system} will be installed onto.
<9> Include this directive to enable FIPS mode on your cluster.
+
[IMPORTANT]
====
If you are configuring nodes to use both disk encryption and mirroring, both features must be configured in the same Butane config.  In addition, if you are configuring disk encryption on a node with FIPS mode enabled, you must include the `fips` directive in the same Butane config, even if FIPS mode is also enabled in a separate manifest.
====

> . In the `openshift` directory, use Butane to create a master or worker manifest from the Butane config. For example, to create a manifest for worker nodes, run the following command:

Reviewer 2: s/master/control plane
